Add CI/CD using GitHub actions guide (#2820)
FusionAuth/fusionauth-site: astro/src/content/docs/get-started/run-in-the-cloud/github-actions.mdx

---
title: GitHub Actions With FusionAuth
description: How to use CI/CD with FusionAuth in GitHub.
section: get started
subcategory: run in the cloud
prerequisites: Docker version 20
url: https://github.com/FusionAuth/fusionauth_github_actions

> codeRoot: https://raw.githubusercontent.com/FusionAuth/fusionauth_github_actions/main

Reviewer: ```suggestion
codeRoot: https://raw.githubusercontent.com/FusionAuth/fusionauth-github-actions/main
```

<AppDatabaseDiagram/>

You might have the following questions:
- How do I deploy these components to production?
- What happens when FusionAuth releases a new version? How do I upgrade FusionAuth without losing data?
- How can I test that login still works after upgrading my app or FusionAuth?
- How do I separate and manage my development and production environments?
- Where do I keep database passwords safely and include them when deploying?

Automating these tasks is called CI/CD or continuous integration and continuous deployment.
- CI refers to testing changes to code you push to the main branch to ensure it fits in with the existing system and everything still works.
- CD refers to deploying every push to the main branch to your production (live) site.

GitHub provides a tool for automating workflows called GitHub Actions. It provides a virtual machine in which you can run scripts to check out your code, compile it, deploy alongside FusionAuth and other services, run tests, and manage deployment.

The flow looks like the diagram below.

<ActionWorkflowDiagram/>

> ## A Simple Example Using GitHub Actions

Reviewer: @rideam I was just thinking about an image that would diagram what sits where because there are so many moving pieces. It is just a suggestion. If you do not think it is worth it, then just resolve this.

![Screenshot 2024-02-22 at 3 55 10 PM](https://github.com/FusionAuth/fusionauth-site/assets/137117976/935b39eb-03ce-4110-b165-332246471357)

The `.github/workflows/test.yaml` file is the GitHub Actions workflow script. The `app/.github/workflows/playwright.yml` file contains a Playwright-provided script, but this script uses more GitHub resources, has a long timeout, and saves the test results to a file. The `test.yaml` script is written to take advantage of the GitHub Free 2000 CI/CD minutes per month.

<Aside type="note">To learn how FusionAuth runs in Docker, please read the [five-minute guide](/docs/quickstarts/5-minute-docker). To learn how Express connects to FusionAuth, please read this [quickstart](/docs/quickstarts/quickstart-javascript-express-web).</Aside>

## How To Upgrade FusionAuth

Before running the app, let's demonstrate how to upgrade FusionAuth. It should usually be as trivial as changing the version number and restarting FusionAuth.

In this section, you will:
- Start an older version of FusionAuth.
- Create a new user.
- Upgrade FusionAuth.
- Check that the new user is still there.

Open the `docker-compose.yml` file. On line 51, set the FusionAuth image to the following.

```yaml
image: fusionauth/fusionauth-app:1.47.1
```

In a terminal, run the following command.

```bash
docker compose up
```

> FusionAuth will start, and you will be able to log in at `http://localhost:9011` with username `admin@example.com` and password `password`. In the Users section of the site, create a new user.

Reviewer: @rideam I was thinking more Users -> Green Add Button or a link on how to add users documentation.

FusionAuth also has a paid version that [runs in the cloud](/docs/get-started/run-in-the-cloud/cloud). If you want to avoid the hassle of managing, backing up, and upgrading FusionAuth yourself, this might be the solution for you.
</Aside>

## Test Your App Login With Playwright

FusionAuth is already running. To start the app that uses it, run the code below.

```bash
cd app
npm install
npm run start
```

Browse to `http://localhost:3000` and log in with the same user as previously.

When changing your application, it would be faster to test that login works automatically, rather than by hand. This is called an integration, or end-to-end test, as opposed to a unit test, which tests only a single function. For integration tests in the browser, you can use Playwright.

<Aside type="note">
Playwright was created by Microsoft, runs in JavaScript, and allows you to run tests in multiple browsers. It is a popular successor to Selenium, which used Java, and Puppeteer, which was written by Google and focused on Chrome.
</Aside>

To start the Playwright tests for this app, open a new terminal and run the code below.

```bash
npx playwright install
npx playwright install-deps

> npx playwright test --project=chromium;

Reviewer: need to add note that the new terminal window needs to be in the app directory.

```

The output should be as below.

```bash
Running 1 test using 1 worker

  ✓  1 [chromium] › test.spec.js:3:1 › Test login (5.1s)

  1 passed (5.8s)
```

> The command returns `0` to the terminal if all tests passed, and another number if any tests failed. This allows you (and GitHub Actions) to write shell scripts that perform different tasks if the tests succeed or not.

Reviewer: I think you should say it just returns '0' if tests pass.  By specifying 'to the terminal', I expect to see it.

To see what is happening visually, run the code below.

```bash
npx playwright test --project=chromium --ui

> ```

Reviewer: I had to click the run button next to the Test to see it run.

Reviewer: This image is not very readable.  May need to just take a section and zoom in so it is more clear.

![Screenshot 2024-02-22 at 4 03 12 PM](https://github.com/FusionAuth/fusionauth-site/assets/137117976/de474f7d-c16c-4b85-b537-78dd0fdd5c1f)

- To add a new test of your own, write a new function like `test('Test login', async ({ page }) => {`.

To add Playwright to your own project, follow [their guide](https://playwright.dev/docs/intro). Running `npm init playwright` will create a configuration file and sample test you can use as a starting point.

In the configuration file `playwright.config.js`, you can set which browsers to use (we used `--project=chromium` earlier), whether to start your app before running the tests, and which directory contains test files.

## Test Your App In A GitHub Action

Now that you know how to test your app after a change to it or FusionAuth, let's automate the process in GitHub. Your goal is to have a test run whenever committing to your main branch. GitHub will email you if the test fails.

Open the file `.github/workflows/test.yaml`. Line 5 currently reads as follows.

```yaml
      - main_RENAME_THIS_TO_ENABLE_TEST
```

Change the line to the following.

```yaml
      - main
```

You've renamed the trigger that causes this action to execute, from a nonsense name to the main branch of your application.

> By default, GitHub prevents action workflows from running on forked repositories. To grant permission for the workflow to execute, navigate to the Actions tab and click on the <InlineUIElement>I understand my workflows, go ahead and enable them</InlineUIElement> button on the repository. Save, commit in Git, and push to GitHub, and the test will execute.

Reviewer: "Save, commit in Git, and push to GitHub, and the test will execute."

I think you mean to change something in your dev environment, commit it and then push to Github.  Or do you want to change something in Github directly? Or does it make a difference.  Not very clear here.

> You can browse to the Actions tab in your GitHub repository to see the test run.

Reviewer: Need more detail on how to see it for new users.  Click on the test and the click on the text window.  That sort of thing.

> 

Reviewer: Getting this error by default.

![Screenshot 2024-02-12 at 4 28 30 PM](https://github.com/FusionAuth/fusionauth-site/assets/137117976/101b33a5-2db9-4746-b0cf-106439ebc57d)

I also got it for 
services.search.environment.bootstrap.memory_lock

I changed the values to string I.E. true -> "true" and it worked.  Not sure if that messes something else up, but needs to be cleaned up in the docker-compose file.

<img src="/img/docs/get-started/run-in-the-cloud/github_actions/passedtest.png" alt="Passed GitHub action test." role="bottom-cropped" />

> To test that the test works, change the email in `test.spec.js` to `wrong@example.com` and push to GitHub again. You can now verify that the test does detect a broken login.

Reviewer: should be more clear:

change:
expect(emailText).toBe('richard@example.com');

to:
expect(emailText).toBe('wrong@example.com');

<img src="/img/docs/get-started/run-in-the-cloud/github_actions/failedtest.png" alt="Failed GitHub action test." role="bottom-cropped" />

If your test runs successfully, you can be certain that your app and FusionAuth are both running, can speak to each other, and that the user is still present in the database after any migrations have run.

## Understand GitHub Actions

> GitHub Actions is a service provided by GitHub that allows you to run workflows that can do anything you can write in a script, triggered by different types of events in your repository. GitHub runs your script in a virtual machine. GitHub allocates each user a certain amount of free CPU minutes per month and a set amount of storage, depending on your plan. For more details, see the [pricing page](https://docs.github.com/en/billing/managing-billing-for-github-actions/about-billing-for-github-actions).

Reviewer: could have link here to Github Actions help

Let's review the action script to see how it works. All scripts are stored in the `.github/workflows` directory, and the files can be called whatever you like. Here, the file is called `test.yaml`.

<RemoteCode url={frontmatter.codeRoot + "/.github/workflows/test.yaml"}
  lang="yaml" />

Each action has a name, which is a human-readable label.

```yaml
name: Test FusionAuth login
```

This is followed by an event, which is all the conditions that trigger the action to execute. Events are stored in the `on` object. The most common ones to use are when pushing a commit or receiving a pull request. For a full list, see the [documentation](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows).

Your action uses the push-to-main-branch event.

```yaml
  push:
    branches:
      - main
```

Then comes the action itself, which is stored in the `jobs` object. This is what will run when the event occurs. Your action has just one job, `run-tests`.

> You need to specify the type of operating system the machine will run to execute the job steps. Here, you use `ubuntu-latest`. It is the cheapest runner on GitHub. Windows runners use minutes at twice the rate of Linux, and Mac runners use ten times the rate of Linux.

Reviewer: Should be in an aside of type note.

<Aside type="note">
Playwright was created by Microsoft, runs in JavaScript, and allows you to run tests in multiple browsers. It is a popular successor to Selenium, which used Java, and Puppeteer, which was written by Google and focused on Chrome.
</Aside>

To start the Playwright tests for this app, open a new terminal in the `app` directory and run the code below.

```bash
npx playwright install
npx playwright install-deps
npx playwright test --project=chromium
```

The output should be as below.

```bash
Running 1 test using 1 worker

  ✓  1 [chromium] › test.spec.js:3:1 › Test login (5.1s)

  1 passed (5.8s)
```

The command returns `0` if all tests passed, and another number if any tests failed. This allows you (and GitHub Actions) to write shell scripts that perform different tasks if the tests succeed or not.

To see what is happening visually, run the code below.

```bash
npx playwright test --project=chromium --ui
```

This will open the playwright window. Click on the `Run` button next to the test name to execute the test. You can see that Playwright runs a browser, clicks buttons, and fills in forms.

> <img src="/img/docs/get-started/run-in-the-cloud/github_actions/playwright.png" alt="Playwright test UI." role="bottom-cropped" />

Reviewer: This image is better, but has a red dot in on the submit button.

Reviewer: makes sense, must have missed that.

## Write A Test For Playwright

> The code for the test Playwright runs is in the file `test.spec.js`.

Reviewer: ```suggestion
The code for the test Playwright runs is in the file `app/tests/test.spec.js`.
```

- To add a new test of your own, write a new function like `test('Test login', async ({ page }) => {`.

To add Playwright to your own project, follow [their guide](https://playwright.dev/docs/intro). Running `npm init playwright` will create a configuration file and sample test you can use as a starting point.

In the configuration file `playwright.config.js`, you can set which browsers to use (we used `--project=chromium` earlier), whether to start your app before running the tests, and which directory contains test files.

## Test Your App In A GitHub Action

Now that you know how to test your app after a change to it or FusionAuth, let's automate the process in GitHub. Your goal is to have a test run whenever committing to your main branch. GitHub will email you if the test fails.

<Aside type="note">
 By default, GitHub prevents action workflows from running on forked repositories. To grant permission for the workflow to execute, navigate to the Actions tab on your forked repository and click on the <InlineUIElement>I understand my workflows, go ahead and enable them</InlineUIElement> button on the repository.
</Aside>

> Open the file `.github/workflows/test.yaml`. Line 5 currently reads as follows.

Reviewer: Need to be more clear here.  Open the file in your IDE? or open the file in the Github website?  I did the website.

```yaml
      - main_RENAME_THIS_TO_ENABLE_TEST
```

Change the line to the following.

```yaml
      - main
```

> You've renamed the trigger that causes this action to execute, from a nonsense name to the main branch of your application. Save the changes to the file, commit in Git, and push to GitHub, and the test will execute.

Reviewer: I see you explain it here, but I think this needs to be moved above. Explain they will change in their branch and then push.

Reviewer: Also worth mention is that the test can take a few minutes to run. Currently 5 minutes for me.

Let's assume your production environment is a virtual server that runs your application in a Go binary and FusionAuth in Docker.

After the tests in your action have run successfully, you can deploy your app and FusionAuth to the server.

To run a command on your server from within an action, use [SSH for GitHub Actions](https://github.com/appleboy/ssh-action). This action allows you to open a terminal on your server and run commands. Here's how you would change the version number of FusionAuth, and restart it in Docker.

```yaml
- name: Update FusionAuth and Restart Docker
  uses: appleboy/ssh-action@master
  with:
    host: ${{ secrets.SERVER_HOST }}
    username: ${{ secrets.SERVER_USERNAME }}
    key: ${{ secrets.SERVER_SSH_KEY }}
    script: |
      cd /fa_config # assume this is where docker compose file is

      # change docker version number
      sed -i 's|fusionauth/fusionauth-app:.*|fusionauth/fusionauth-app:1.48.3|' docker-compose.yml

      # pull the latest image from DockerHub before we try to use it
      docker-compose pull fusionauth-app

      # restart docker and leave it running
      docker-compose up -d fusionauth-app
```

> We'll discuss the secrets in a later section.

Reviewer: Move this above the example.

> To deploy your Go application, you can do one of the following:

Reviewer: ```suggestion
Let's say you want to deploy a Go application you created. You can do one of the following:
```

Once you have your application files ready, you can SSH into your server using the SSH action, download the Docker image or binary files, run any database migrations scripts, and restart your web application.

## Test Your Production Release

Even if your tests run successfully in the GitHub virtual machine, your production release might not allow users to log in. You need to test again.

Add a final step to your script, after the deployment step, that runs the Playwright login test. This time, run a test file that uses the production server URL instead of your localhost in the line `await page.goto('http://localhost:3000');`.

If this test passes, you can be certain your live website and FusionAuth can still communicate. If it fails, you need to revert your release to the previous version.

## Secrets And Deployment Environments

Your organization will probably have several environments (server and database) where your application runs:
Environment | Purpose
--- | ---

> Development | Where programmers are free to test new features without fear of breaking anything. Database users are anonymized.

Reviewer: vale does not like anonymized.

Test | Where quality assurance testers test the latest release that is scheduled to go live soon.

> Preproduction | An exact duplicate of the production environment, with real sensitive user information. This is used for debugging errors in production. Access is restricted.

Reviewer: ```suggestion
Pre-production | An exact duplicate of the production environment, with real sensitive user information. This is used for debugging errors in production. Access is restricted.
```
